// @kolkov/angular-editor/esm2015/lib/config.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/config.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function CustomClass() { }
if (false) {
    /** @type {?} */
    CustomClass.prototype.name;
    /** @type {?} */
    CustomClass.prototype.class;
    /** @type {?|undefined} */
    CustomClass.prototype.tag;
}
/**
 * @record
 */
export function Font() { }
if (false) {
    /** @type {?} */
    Font.prototype.name;
    /** @type {?} */
    Font.prototype.class;
}
/**
 * @record
 */
export function AngularEditorConfig() { }
if (false) {
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.editable;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.spellcheck;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.height;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.minHeight;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.maxHeight;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.width;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.minWidth;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.translate;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.enableToolbar;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.showToolbar;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.placeholder;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.defaultParagraphSeparator;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.defaultFontName;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.defaultFontSize;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.uploadUrl;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.uploadWithCredentials;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.fonts;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.customClasses;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.sanitize;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.toolbarPosition;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.outline;
    /** @type {?|undefined} */
    AngularEditorConfig.prototype.toolbarHiddenButtons;
}
/** @type {?} */
export const angularEditorConfig = {
    editable: true,
    spellcheck: true,
    height: 'auto',
    minHeight: '0',
    maxHeight: 'auto',
    width: 'auto',
    minWidth: '0',
    translate: 'yes',
    enableToolbar: true,
    showToolbar: true,
    placeholder: 'Enter text here...',
    defaultParagraphSeparator: '',
    defaultFontName: '',
    defaultFontSize: '',
    fonts: [
        { class: 'arial', name: 'Arial' },
        { class: 'times-new-roman', name: 'Times New Roman' },
        { class: 'calibri', name: 'Calibri' },
        { class: 'comic-sans-ms', name: 'Comic Sans MS' }
    ],
    uploadUrl: 'v1/image',
    uploadWithCredentials: false,
    sanitize: true,
    toolbarPosition: 'top',
    outline: true,
};
